File: hackerrank/problem solving/data structures/1 - arrays/5 - sparse arrays/sparse_arrays.go
// https://www.hackerrank.com/challenges/sparse-arrays/problem?isFullScreen=true

package main

import "fmt"

func main() {
    var n int
    _, _ = fmt.Scan(&n)
    strings := readLines(n)
    stringsToCounts := counter(strings)

    _, _ = fmt.Scan(&n)
    queries := readLines(n)

    for _, x := range countsFrom(stringsToCounts, queries) {
        fmt.Println(x)
    }
}

func readLines(n int) []string {
    arr := make([]string, n)
    for i := range n {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

func counter(arr []string) map[string]int {
    freqMap := map[string]int{}
    for _, x := range arr {
        freqMap[x]++
    }
    return freqMap
}

// n: length of queries
// maxQueryStringLength: 20
// T: O(n * maxQueryStringLength) = O(n * 20) = O(n)
// S: O(n) extra space
func countsFrom(freqMap map[string]int, queries []string) []int {
    arr := make([]int, len(queries))
    for i, q := range queries {
        arr[i] = freqMap[q]
    }
    return arr
}

File: hackerrank/problem solving/data structures/1 - arrays/5 - sparse arrays/sparse_arrays.java
// https://www.hackerrank.com/challenges/sparse-arrays/problem?isFullScreen=true
// Java 25

void main() {
    int n = Integer.parseInt(IO.readln());
    String[] strings = readLines(n);
    Map<String, Integer> stringsToCounts = counter(strings);

    n = Integer.parseInt(IO.readln());
    String[] queries = readLines(n);

    for (var x : countsFrom(stringsToCounts, queries))
        IO.println(x);
}

String[] readLines(int n) {
    return IntStream.range(0, n).mapToObj(_ -> IO.readln()).toArray(String[]::new);
}

Map<String, Integer> counter(String[] arr) {
    var freqMap = new HashMap<String, Integer>();
    for (var x : arr)
        freqMap.put(x, freqMap.getOrDefault(x, 0) + 1);
    return freqMap;
}

// n: length of queries
// maxQueryStringLength: 20
// T: O(n * maxQueryStringLength) = O(n * 20) = O(n)
// S: O(n) extra space
int[] countsFrom(Map<String, Integer> freqMap, String[] queries) {
    return Arrays.stream(queries).mapToInt(q -> freqMap.getOrDefault(q, 0)).toArray();
}
